staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply.go

```diff
@@ -0,0 +1,180 @@
+	patch := make(map[string]interface{})
+	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
+	}
+	return patch, nil
+}
+
+func (p *applyPatcher) convertResultToUnversioned(result apply.Result) (runtime.Object, error) {
+	voutput, err := p.creater.New(p.kind)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create empty output object: %v", err)
+	}
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.MergedResult.(map[string]interface{}), voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result back: %v", err)
+	}
+	p.defaulter.Default(voutput)
+
+	gvk := p.kind.GroupKind().WithVersion(runtime.APIVersionInternal)
+	uoutput, err := p.unsafeConvertor.ConvertToVersion(voutput, gvk.GroupVersion())
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to unversioned: %v", err)
+	}
+	return uoutput, nil
+}
+
+func (p *applyPatcher) saveNewIntent(patch map[string]interface{}, workflow string, dst runtime.Object) error {
+	// Make sure we have the gvk set on the object.
+	(&unstructured.Unstructured{Object: patch}).SetGroupVersionKind(p.kind)
+
+	j, err := json.Marshal(patch)
+	if err != nil {
+		return fmt.Errorf("failed to serialize json: %v", err)
+	}
+
+	accessor, err := meta.Accessor(dst)
+	if err != nil {
+		return fmt.Errorf("couldn't get accessor: %v", err)
+	}
+	m := accessor.GetManagedFields()
+	if m == nil {
+		m = make(map[string]metav1.VersionedFieldSet)
+	}
+	// TODO: save the managedFields correctly
+	m[workflow] = metav1.VersionedFieldSet{
+		APIVersion: string(j),
+	}
+	accessor.SetManagedFields(m)
+	return nil
+}
+
+func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
+	current, err := p.convertCurrentVersion(currentObject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert current object: %v", err)
+	}
+
+	lastIntent, err := p.extractLastIntent(currentObject, workflowId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract last intent: %v", err)
+	}
+	newIntent, err := p.getNewIntent()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get new intent: %v", err)
+	}
+
+	element, err := parse.CreateElement(lastIntent, newIntent, current, p.model)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse elements: %v", err)
+	}
+	result, err := element.Merge(strategy.Create(strategy.Options{}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge elements: %v", err)
+	}
+
+	output, err := p.convertResultToUnversioned(result)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result: %v", err)
+	}
+
+	if err := p.saveNewIntent(newIntent, workflowId, output); err != nil {
+		return nil, fmt.Errorf("failed to save last intent: %v", err)
+	}
+
+	// TODO(apelisse): Check for conflicts with other managedFields
+	// and report actionable errors to users.
```

Comment:
What are the semantics here?  Do we have a meta-mapping that e.g. v1's `podCIDR` -> v2's `podCIDRs` ?